pkg/cableengine/syncgatewaystatus.go

```diff
@@ -0,0 +1,161 @@
+
+	i.Lock()
+	defer i.Unlock()
+
+	klog.V(log.TRACE).Info("Running syncGatewayStatus()")
+
+	gatewayObj, err := i.generateGatewayObject()
+	if err != nil {
+		klog.Errorf("error generating gateway object from driver connections: %s", err)
+		return
+	}
+
+	existingGw, err := i.getLastSyncedGateway(gatewayObj.Name)
+
+	// log and stop for any error different to a not-found error
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("error trying to read existing gateway from k8s api: %s", err)
+		return
+	}
+
+	gwClient := i.clientset.SubmarinerV1().Gateways(i.namespace)
+
+	if err != nil && errors.IsNotFound(err) {
+		klog.V(log.TRACE).Infof("Gateway object needs creation: %+v", gatewayObj)
+		_, err = gwClient.Create(gatewayObj)
+		if err != nil {
+			klog.Errorf("error creating Gateway object: %s", err)
+			return
+		}
+	} else {
+		if !reflect.DeepEqual(gatewayObj.Status, existingGw.Status) {
+			klog.V(log.TRACE).Infof("Gateway object needs an update: %+v", gatewayObj)
+			existingGw.Status = gatewayObj.Status
+			existingGw.Labels = gatewayObj.Labels
+			existingGw.Annotations = gatewayObj.Annotations
+
+			gw, err := gwClient.Update(existingGw)
+			if err != nil {
+				klog.Errorf("error updating Gateway object: %s", err)
+				return
+			} else {
+				klog.V(log.TRACE).Infof("Gateway updated correctly: %+v", gw)
+			}
+		} else {
+			klog.V(log.TRACE).Info("Gateway object didn't need an update")
+		}
+	}
+
+	if gatewayObj.Status.HAStatus == v1.HAStatusActive {
+		err := i.cleanupStaleGatewayEntries()
+		if err != nil {
+			klog.Errorf("error cleaning up stale gateway entries: %s", err)
+		}
+	}
+}
+
+func (i *engine) cleanupStaleGatewayEntries() error {
+	gwClient := i.clientset.SubmarinerV1().Gateways(i.namespace)
+	gateways, err := gwClient.List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, gw := range gateways.Items {
+		stale, err := isGatewayStale(gw)
+		if err != nil {
+			// In this case we don't want to stop the cleanup loop and just log it
+			klog.Errorf("error processing stale gateway: %+v , %s", gw, err)
+		}
+		if stale {
+			err := gwClient.Delete(gw.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				// In this case we don't want to stop the cleanup loop and just log it
+				klog.Errorf("error deleting stale gateway: %+v, %s", gw, err)
+			}
+			klog.Warningf("deleted stale gateway: %s, didn't report for %d seconds",
+				gw.Name, GatewayStaleTimeoutSeconds)
+		}
+	}
+	return nil
+}
+
+func isGatewayStale(gateway v1.Gateway) (bool, error) {
+
+	timestamp, ok := gateway.ObjectMeta.Annotations[updateTimestampAnnotation]
+	if !ok {
+		return true, fmt.Errorf("%q annotation not found", updateTimestampAnnotation)
+	}
+
+	timestampInt, err := strconv.ParseInt(timestamp, 10, 64)
+	if err != nil {
+		return true, fmt.Errorf("error parsing update-timestamp: %s", err)
+	}
+	now := time.Now().UTC().Unix()
+
+	return now >= (timestampInt + GatewayStaleTimeoutSeconds), nil
```

Comment:
No, it's a pre-requisite of a running k8s cluster to be NTP time synced, and we use UTC which does not depend of DST time changes.

It's a foundation of any clustered service, and a common methodology to use timestamps.

see https://docs.oracle.com/en/operating-systems/oracle-linux/kubernetes/section_q4v_phr_p2b_ha.html

This kind of stale consideration was also used in openstack-neutron for a long time, still being used.

I'd prefer to avoid the extra complexity while we can.

Reply:
Also see ETCD: https://github.com/etcd-io/etcd/issues/9768   also depends on this fact.

Reply:
makes sense.